File: contact/src/main/java/com/sch/fakecontacts/ui/activities/MainActivity.java
package com.sch.fakecontacts.ui.activities;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import android.support.v4.app.FragmentActivity;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.SwitchCompat;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.sch.fakecontacts.R;
import com.sch.fakecontacts.model.generator.ContactGenerator;
import com.sch.fakecontacts.model.generator.GenerationOptions;
import com.sch.fakecontacts.ui.dialogs.ProgressDialogController;

import permissions.dispatcher.NeedsPermission;
import permissions.dispatcher.RuntimePermissions;

@RuntimePermissions
public class MainActivity extends AppCompatActivity implements View.OnClickListener {
    private static final int REQUEST_CODE_SELECT_GROUP = 0;

    private static final String PREF_CONTACT_COUNT = "contact_count";
    private static final String PREF_WITH_EMAILS = "with_emails";
    private static final String PREF_WITH_PHONES = "with_phones";
    private static final String PREF_WITH_ADDRESSES = "with_addresses";
    private static final String PREF_WITH_AVATARS = "with_avatars";
    private static final String PREF_WITH_EVENTS = "with_events";
    private static final String PREF_OVERWRITE_EXISTING = "overwrite_existing";

    private SwitchCompat withEmailsView;
    private SwitchCompat withPhonesView;
    private SwitchCompat withAddressesView;
    private SwitchCompat withAvatarsView;
    private SwitchCompat withEventsView;
    private SwitchCompat overwriteExistingView;
    private EditText countView;
    private Button selectGroupButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        countView = (EditText) findViewById(R.id.edit_count);
        withEmailsView = (SwitchCompat) findViewById(R.id.switch_with_emails);
        withPhonesView = (SwitchCompat) findViewById(R.id.switch_with_phones);
        withAddressesView = (SwitchCompat) findViewById(R.id.switch_with_addresses);
        withAvatarsView = (SwitchCompat) findViewById(R.id.switch_with_avatars);
        withEventsView = (SwitchCompat) findViewById(R.id.switch_with_events);
        overwriteExistingView = (SwitchCompat) findViewById(R.id.switch_overwrite_existing);

        final Button generateButton = (Button) findViewById(R.id.button_generate);
        generateButton.setOnClickListener(this);

        selectGroupButton = (Button) findViewById(R.id.button_select_group);
        selectGroupButton.setOnClickListener(this);

        restorePersistentState();
    }

    @Override
    protected void onStop() {
        super.onStop();
        savePersistentState();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int i = item.getItemId();
        if (i == R.id.action_open_contacts) {
            openContactsApp();
            return true;
        } else if (i == R.id.action_delete_generated_contacts) {
            MainActivityPermissionsDispatcher.deleteGeneratedContactsWithCheck(this);
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        switch (requestCode) {
            case REQUEST_CODE_SELECT_GROUP:
                if (resultCode == RESULT_OK) {
                    final long groupId = data.getLongExtra(GroupsActivity.EXTRA_GROUP_ID, 0);
                    final String groupName = data.getStringExtra(GroupsActivity.EXTRA_GROUP_NAME);
                    selectGroupButton.setText(groupName);
                    selectGroupButton.setTag(groupId);
                }
                break;
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        MainActivityPermissionsDispatcher.onRequestPermissionsResult(this, requestCode, grantResults);
    }

    @Override
    public void onClick(View v) {
        int i = v.getId();
        if (i == R.id.button_select_group) {
            MainActivityPermissionsDispatcher.selectGroupWithCheck(this);

        } else if (i == R.id.button_generate) {
            if (validateParams()) {
                MainActivityPermissionsDispatcher.generateContactsWithCheck(this);
            }

        }
    }

    private void openContactsApp() {
        final Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse("content://contacts/people/"));
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        startActivity(intent);
    }

    private int getContactCount() {
        return countView.length() > 0 ? Integer.parseInt(countView.getText().toString()) : 0;
    }

    private boolean validateParams() {
        return countView.length() > 0;
    }

    @NeedsPermission(Manifest.permission.READ_CONTACTS)
    void selectGroup() {
        GroupsActivity.startForResult(this, REQUEST_CODE_SELECT_GROUP);
    }

    @NeedsPermission(Manifest.permission.WRITE_CONTACTS)
    void generateContacts() {
        final Long groupId = (Long) selectGroupButton.getTag();
        final GenerationOptions.Builder builder = new GenerationOptions.Builder()
                .setContactCount(getContactCount())
                .setOverwriteExisting(overwriteExistingView.isChecked());

        if (groupId != null) {
            builder.setGroupId(groupId);
        }
        if (withEmailsView.isChecked()) {
            builder.withEmails();
        }
        if (withPhonesView.isChecked()) {
            builder.withPhones();
        }
        if (withAddressesView.isChecked()) {
            builder.withAddresses();
        }
        if (withAvatarsView.isChecked()) {
            builder.withAvatars();
        }
        if (withEventsView.isChecked()) {
            builder.withEvents();
        }
        new GenerateContactsTask(this, builder.build()).execute();
    }

    @NeedsPermission(Manifest.permission.WRITE_CONTACTS)
    void deleteGeneratedContacts() {
        new DeleteGeneratedContactsTask(this).execute();
    }

    private void restorePersistentState() {
        final int contactCount = getPreferences().getInt(PREF_CONTACT_COUNT, 100);
        countView.setText(String.valueOf(contactCount));

        withEmailsView.setChecked(getPreferences().getBoolean(PREF_WITH_EMAILS, true));
        withPhonesView.setChecked(getPreferences().getBoolean(PREF_WITH_PHONES, true));
        withAddressesView.setChecked(getPreferences().getBoolean(PREF_WITH_ADDRESSES, false));
        withAvatarsView.setChecked(getPreferences().getBoolean(PREF_WITH_AVATARS, false));
        withEventsView.setChecked(getPreferences().getBoolean(PREF_WITH_EVENTS, true));

        overwriteExistingView.setChecked(getPreferences().getBoolean(PREF_OVERWRITE_EXISTING, true));
    }

    private void savePersistentState() {
        getPreferences().edit()
                .putInt(PREF_CONTACT_COUNT, getContactCount())
                .putBoolean(PREF_WITH_EMAILS, withEmailsView.isChecked())
                .putBoolean(PREF_WITH_PHONES, withPhonesView.isChecked())
                .putBoolean(PREF_WITH_ADDRESSES, withAddressesView.isChecked())
                .putBoolean(PREF_WITH_AVATARS, withAvatarsView.isChecked())
                .putBoolean(PREF_WITH_EVENTS, withEventsView.isChecked())
                .putBoolean(PREF_OVERWRITE_EXISTING, overwriteExistingView.isChecked())
                .apply();
    }

    private SharedPreferences getPreferences() {
        return PreferenceManager.getDefaultSharedPreferences(this);
    }

    private static class GenerateContactsTask extends AsyncTask<Void, Void, Void> {
        private final Context context;
        private final GenerationOptions options;
        private final ProgressDialogController progressDialogController;

        GenerateContactsTask(FragmentActivity activity, GenerationOptions options) {
            this.context = activity;
            this.options = options;
            progressDialogController = new ProgressDialogController(activity, activity.getSupportFragmentManager());
        }

        @Override
        protected void onPreExecute() {
            super.onPreExecute();
            progressDialogController.show(R.string.message_please_wait);
        }

        @Override
        protected Void doInBackground(Void... params) {
            new ContactGenerator(context).generate(options);
            return null;
        }

        @Override
        protected void onPostExecute(Void result) {
            super.onPostExecute(result);
            progressDialogController.dismiss();
            Toast.makeText(context, R.string.message_done, Toast.LENGTH_SHORT).show();
        }
    }

    private static class DeleteGeneratedContactsTask extends AsyncTask<Void, Void, Void> {
        private final Context context;
        private final ProgressDialogController progressDialogController;

        DeleteGeneratedContactsTask(FragmentActivity activity) {
            context = activity;
            progressDialogController = new ProgressDialogController(activity, activity.getSupportFragmentManager());
        }

        @Override
        protected void onPreExecute() {
            super.onPreExecute();
            progressDialogController.show(R.string.message_please_wait);
        }

        @Override
        protected Void doInBackground(Void... voids) {
            new ContactGenerator(context).deleteGeneratedContacts();
            return null;
        }

        @Override
        protected void onPostExecute(Void aVoid) {
            super.onPostExecute(aVoid);
            progressDialogController.dismiss();
            Toast.makeText(context, R.string.message_done, Toast.LENGTH_SHORT).show();
        }
    }
}

File: app/src/main/java/org/gosky/fakeinfo/MainActivity.kt
package org.gosky.fakeinfo

import android.Manifest
import android.annotation.TargetApi
import android.content.ContentValues
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.AsyncTask
import android.os.Build
import android.os.Bundle
import android.provider.CallLog
import android.provider.Telephony
import android.support.v4.app.FragmentActivity
import android.support.v7.app.AppCompatActivity
import android.widget.Toast
import com.sch.fakecontacts.model.generator.ContactGenerator
import com.sch.fakecontacts.model.generator.GenerationOptions
import com.sch.fakecontacts.ui.dialogs.ProgressDialogController
import com.tbruyelle.rxpermissions.RxPermissions
import kotlinx.android.synthetic.main.activity_main.*
import java.text.ParseException
import java.util.*

class MainActivity : AppCompatActivity() {
    val callType = arrayOf(CallLog.Calls.INCOMING_TYPE,
            CallLog.Calls.OUTGOING_TYPE, CallLog.Calls.MISSED_TYPE
    )


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        btn_chuowo.setOnClickListener {
            RxPermissions.getInstance(this)
                    .request(Manifest.permission.WRITE_CONTACTS, Manifest.permission.WRITE_CALL_LOG)
                    .subscribe {
                        generateContacts()
                    }
        }
        if (Build.VERSION.SDK_INT >= 20) {
            if (packageName != getSystemDefaultSms()) {
                setDefaultSms(packageName)
            }
        }

    }


    fun generateContacts() {
        val builder = GenerationOptions.Builder()
                .setContactCount(et_contact_count.text.toString().toInt())
                .setOverwriteExisting(true)
//        builder.withEmails()
        builder.withPhones()
//        builder.withAddresses()
//        builder.withAvatars()
//        builder.withEvents()
        GenerateContactsTask(this, builder.build()).execute()
    }

    fun generateCallLogs() {
        val cv = ContentValues()
        cv.put(CallLog.Calls.NUMBER, getRandomContact())
        cv.put(CallLog.Calls.DURATION, Random().nextInt(60))
        cv.put(CallLog.Calls.NEW, 1)
        cv.put(CallLog.Calls.DATE, System.currentTimeMillis() - Random().nextInt(1000 * 1000))
        cv.put(CallLog.Calls.TYPE, callType[Random().nextInt(callType.size)])
        cv.put(CallLog.Calls.CACHED_NAME, "")
        cv.put(CallLog.Calls.CACHED_NUMBER_TYPE, 0)
        cv.put(CallLog.Calls.CACHED_NUMBER_LABEL, "")
        contentResolver.insert(CallLog.Calls.CONTENT_URI, cv)
    }


    fun generateSms() {

        if (Build.VERSION.SDK_INT >= 20) {
            if (packageName != getSystemDefaultSms()) {
                setDefaultSms(packageName)
            }
        }
        val values = ContentValues()

        try {
            values.put("date", System.currentTimeMillis() - Random().nextInt(1000 * 1000))
            values.put("address", getRandomContact())
            values.put("body", GeneratorSmsContent.random())
            values.put("type", "2")
            values.put("read", "1")//"1"means has read ,1表示已读

            contentResolver.insert(Uri.parse("content://sms/inbox"), values)
//            Toast.makeText(this@MainActivity, "短信插入成功，部分手机的收件箱有延迟，请等候", Toast.LENGTH_SHORT).show()

        } catch (e: ParseException) {
            Toast.makeText(this@MainActivity, "时间输入异常，请重新尝试", Toast.LENGTH_SHORT).show()
            e.printStackTrace()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun getRandomContact(): String? {
        if (Random().nextInt(100) > 50) {
            val list = ContactUtils.getAllContacts(this).entries.map { it.key }
            if (list.isNotEmpty())
                return list[Random().nextInt(list.size)]
        }
        return getRandomTelNumber()
    }


    @TargetApi(Build.VERSION_CODES.KITKAT)
    fun getSystemDefaultSms(): String {
        return Telephony.Sms.getDefaultSmsPackage(this)
    }

    fun setDefaultSms(packageName: String) {
        val intent = Intent(Telephony.Sms.Intents.ACTION_CHANGE_DEFAULT)
        intent.putExtra(Telephony.Sms.Intents.EXTRA_PACKAGE_NAME, packageName)
        startActivity(intent)
    }

    private inner class GenerateContactsTask constructor(activity: FragmentActivity, private val options: GenerationOptions) : AsyncTask<Void, Void, Void>() {
        private val context: Context
        private val progressDialogController: ProgressDialogController

        init {
            this.context = activity
            progressDialogController = ProgressDialogController(activity, activity.supportFragmentManager)
        }

        override fun onPreExecute() {
            super.onPreExecute()
            progressDialogController.show(com.sch.fakecontacts.R.string.message_please_wait)
        }

        override fun doInBackground(vararg params: Void): Void? {
            ContactGenerator(context).generate(options)
            for (i in 1..et_call_logs_count.text.toString().toInt()) {
                generateCallLogs()
            }
            for (i in 1..et_sms_count.text.toString().toInt()) {
                generateSms()
            }
            return null
        }

        override fun onPostExecute(result: Void?) {
            super.onPostExecute(result)
            progressDialogController.dismiss()
            Toast.makeText(context, com.sch.fakecontacts.R.string.message_done, Toast.LENGTH_SHORT).show()
        }
    }
}
